### ISSUE TITLE
ConvertFrom-String is missing

### ISSUE THREAD

**Issue by vors** [PowerShell/PowerShell#726]
FlashExtact is MSR project.
@LeeHolmes originally took code of the project instead of nuget dependencies, because of the Windows build requirements.

We should replace FlashExtract sources by Nuget packages and bring the code for `ConvertFrom-String`.

**Comment by iSazonov** [comment]
> Home page: https://microsoft.github.io/prose/
> Nuget packages: https://www.nuget.org/packages/Microsoft.ProgramSynthesis/
It seems this project for a long time had no updates. Is it frozen?

**Comment by iSazonov** [comment]
I see now ProgramSynthesis lives! Current version is 4.0.0.
Can we get any progress with the Issue?

**Comment by SteveL-MSFT** [comment]
Don't think we'll get to this for 6.0.0, putting it in 6.1.0 for now

**Comment by AceHack** [comment]
This is a really important feature, I am sad to see it moved to 6.2

**Comment by SteveL-MSFT** [comment]
@AceHack we need an OSS implementation of the FlashExtract engine before we can port this to PSCore6

**Comment by SteveL-MSFT** [comment]
Looks like https://www.nuget.org/packages/Microsoft.ProgramSynthesis/ has been getting regular updates, so this is viable.

**Comment by iSazonov** [comment]
@SteveL-MSFT Can MSFT share again the code of Convert-String and Convert-From-String that was removed from the repo?

**Comment by SteveL-MSFT** [comment]
@iSazonov Let me see if I can add it back, the problem is making sure to remove any use of private APIs...

**Comment by SteveL-MSFT** [comment]
@iSazonov I took a look at this.  The Windows PowerShell code for `ConvertFrom-String` and `Convert-String` is 90% proprietary code with 10% being the cmdlet code that calls a wrapper class to do the actual work.  It may be easier to just create a new set of cmdlets for similar functionality.

**Comment by vexx32** [comment]
I'm not sure that's really worth the effort. In every case where I've tried to use that cmdlet for anything more than the simplest tasks, I've ended up giving up after a good hour or so and just using regex instead.

A simple cmdlet to generate PSCustomObjects from strings via a regex pattern and match groups would be far more useful (if pretty straightforward) in my opinion.

**Comment by iSazonov** [comment]
@SteveL-MSFT Please clarify - can we re-implement the cmdlets (with same names) or we need create new cmdlets (with new names)? For first case we need you share Pester tests to ensure compatibility.

**Comment by SteveL-MSFT** [comment]
@iSazonov It's not clear to me how many people actually used those cmdlets that we would need to retain syntax.  I've always found them difficult to use with inconsistent results, so I didn't use them much myself.  My suggestion would be to start with understanding the use cases and optimizing for that rather than focusing on compatibility.  Looking in the old code base, there isn't much tests besides some basic ones:

```powershell
Describe "DelimitedText" -tags 'DRT' {

    It "verifies automatic property generation" {
            
        $result = "Hello 9", "Hello 10", "Hello 90" | ConvertFrom-String
        
        ## Verify first properties got extracted
        $result[0].P1 | Should be 'Hello'
        $result[1].P1 | Should be 'Hello'
        $result[2].P1 | Should be 'Hello'
        
        ## Verify second properties got extracted
        $result[0].P2 | Should be 9
        $result[1].P2 | Should be 10
        $result[2].P2 | Should be 90
	}

    It "verifies property overflow generation" {
            
        $result = "Hello 9" | ConvertFrom-String -PropertyNames A
        
        $result.A | Should be 'Hello'
        $result.P2 | Should be 9
	}

    It "verifies property renaming" {
            
        $result = "Hello 9" | ConvertFrom-String -PN B,C
        
        $result.B | Should be 'Hello'
        $result.C | Should be '9'
	}

    It "verifies property typing of numbers" {
            
        $result = "Hello 9" | ConvertFrom-String -Property B,C
        $result.C.GetType().FullName | Should be 'System.Byte'
	}
    
    It "verifies property typing of TimeSpan" {
            
        $result = "Hello 1:00" | ConvertFrom-String -Property B,C
        $result.C.GetType().FullName | Should be 'System.TimeSpan'
	}

    It "verifies property typing of DateTime" {
            
        $result = "Hello 1/1/2012" | ConvertFrom-String -Property B,C
        $result.C.GetType().FullName | Should be 'System.DateTime'
	}

    It "verifies property typing of Char" {
            
        $result = "Hello A" | ConvertFrom-String -Property B,C
        $result.C.GetType().FullName | Should be 'System.Char'
	}
    
    It "verifies empty strings don't turn into INTs" {
            
        $result = "Hello" | ConvertFrom-String -Delimiter 'l'
        $result.P2.GetType().FullName | Should be 'System.String'
	}   

    It "verifies property typing of String" {
            
        $result = "Hello World" | ConvertFrom-String -Property B,C
        $result.C.GetType().FullName | Should be 'System.String'
	}
    
    It "verifies the ability to change the delimiter" {
            
        $result = "Hello-World" | ConvertFrom-String -Delimiter '-'
        $result.P1 | Should be 'Hello'
        $result.P2 | Should be 'World'
	}
    
    It "verifies that only matching text gets parsed" {
            
        $result = "Foo1","Hello1 World1","Hello-World" | ConvertFrom-String -Delimiter '-'
        $result.P1 | Should be 'Hello'
        $result.P2 | Should be 'World'
        @($result).Count | Should be 1
	}
    
    It "verifies that a good error message is returned from an invalid regular expression" {
            
        try
        {
            $result = "Hello World" | ConvertFrom-String -Delimiter '['
        }
        catch
        {
            $errorRecord = $_
        }
        
        $errorRecord.FullyQualifiedErrorId | Should be "InvalidRegularExpression,Microsoft.PowerShell.Commands.StringManipulation.ConvertFromStringCommand"
	}   
}

$here = Split-Path -Parent $MyInvocation.MyCommand.Path
$data = Join-Path $here "StringManipulationData\ConvertString"

Describe "Convert-String test cases" -tags 'DRT' {
	
	It "Changes first and last name with one Example" {
        
        $result = "Gustavo Soares" | Convert-String -Example "camilla araujo=araujo, c."
        $result | Should be "Soares, G."
    }   

	It "Changes first and last name with one Example, and three inputs" {
        
        $result = "Lee Holmes", "Gustavo Soares", "Sumit Gulwani", "Vu Le" |
            Convert-String -Example "camilla araujo=araujo, c."
        
        $result[0] | Should be "Holmes, L."
        $result[1] | Should be "Soares, G."
		$result[2] | Should be "Gulwani, S."
		$result[3] | Should be "Le, V."
    }	
			
    It "Changes first and last name with two Examples" {
        
        $examples = [PSCustomObject] @{ Before = 'camilla araujo'; After = 'araujo, c.' },
            [PSCustomObject] @{ Before = 'lee holmes'; After = 'holmes, l.' }
        $result = "Gustavo Soares" | Convert-String -Example $examples
        
        $result | Should be "Soares, G."
    }
	
	It "Changes first and last name with one dictionary example" {
		$result = "Gustavo Soares" | Convert-String -Example @{ Before = "camilla araujo"; After = "araujo, c." }
	}
	
	It "Changes first and last name with two dictionary example" {
		$result = "Gustavo Soares" | Convert-String -Example @(@{ Before = "camilla araujo"; After = "araujo, c." },@{ Before = "vu le"; After = "le, v." })
	}

	It "Check invalid text example" {    
		{ "Gustavo Soares" | Convert-String -Example "camilla araujo" } | Should Throw
    }

	It "Check invalid psobject examples" {
        $examples = Import-Csv $data\incorrect-examples.csv
        { "Gustavo Soares" | Convert-String -Example $examples } | Should Throw        
    }	

	It "Replace by empty" {
        $examples = Import-Csv $data\replace-name-by-empty.csv
		$result = "Gustavo Soares" | Convert-String -Example $examples
        $result.length -eq 0 | Should be true        
    }	
}
```

**Comment by iSazonov** [comment]
@SteveL-MSFT Thanks! 
> My suggestion would be to start with understanding the use cases

I agree. That the two cmdlets look like a small demo. [PROSE home site](https://microsoft.github.io/prose/) has more samples. Parsing log, trace, tabular files, JSON, YAML.
PROSE implements still many great APIs.

**Comment by rkeithhill** [comment]
> I've always found them difficult to use with inconsistent results

Unfortunately, that has been my experience as well.

**Comment by iSazonov** [comment]
What do you have in mind? Remove or improve?

**Comment by AceHack** [comment]
I would really like to see this or similar functionality in core.

**Comment by iSazonov** [comment]
The question is how to do it more conveniently.

**Comment by MartinHatch** [comment]
Am I missing something or don't ConvertTo and ConvertFrom essentially solve the same problem?

e.g.
$csv = ConvertFrom-String $stringValue -Format CSV
# or
$csv = ConvertTo-CSV $stringValue

I'm also not really that keen about a "ConvertFrom-String" given how common a string input can be as it ends up being a swiss-army-knife function .. it has to support a potentially massive number of potential output formats / types.

My initial thoughts (although I really haven't analysed this in detail) from a code management/quality perspective to have separate parsing functions from each object.. so it is clear which objects support parsing from a string (and other formats), and which ones do not. 

Just my $0.02 - but would be genuinely interested why it is thought "ConvertFrom" is a better approach?

**Comment by KirkMunro** [comment]
As a proxy solves this problem, but only for Windows. I just came looking for this because I wanted to look at the source code, since it could use some improvements.

I agree that this cmdlet can be difficult to use; however, it is also of significant value when you are parsing string data, which you end up doing a lot with native executables, of which there are plenty these days on Windows, Linux, and macOS. It is much easier than using `regex` to do the same.

Anyhow, I wanted to voice my opinion that I think this functionality should be brought forward in an open source command so that we can enhance it without reinventing the wheel.

**Comment by iSazonov** [comment]
>  I wanted to look at the source code

It is in `source-deport` git branch.

> I agree that this cmdlet can be difficult to use

Perhaps we could design some helper cmdlets.

**Comment by KirkMunro** [comment]
> It is in `source-deport` git branch.

Link? I browsed to that branch, the only thing a search for `ConvertFromStringCommand` turns up is this issue.

**Comment by iSazonov** [comment]
:-) You could look history before https://github.com/PowerShell/PowerShell/commit/0db07f84354ef661f2c1a33676e6346248a40e16

I should say that this will help little because the PROSE engine has been completely rewritten.

**Comment by iRon7** [comment]
Besides the purposed #16257 [`String` >>>Regex>>> `PSCustomObject`](https://github.com/PowerShell/PowerShell/issues/16257), I just created a new request #16677 [Add table support to ConvertFrom-String](https://github.com/PowerShell/PowerShell/issues/16677) for a possible additional `ConvertFrom-String` feature.

**Comment by iRon7** [comment]
> It may be easier to just create a new set of cmdlets for similar functionality.

If it comes to a new set/cmdlet, `ConvertFrom-Text` might be a proper name.

**Comment by jdhitsolutions** [comment]
Under review by the Cmdlet Working Group.

**Comment by iRon7** [comment]
> It is not possible to release an open-source version of this command in PowerShell 7

This brings me to two legal questions:
* Is it allowed to use the same cmdlet name (`ConvertFrom-String`) for a new cmdlet?
* Is it allowed to build a wrapper around the `ConvertFrom-String` cmdlet (without changing the actual cmdlet)?

I guess that there aren't any restrictions in reusing the cmdlet name (otherwise cmdlet names should be registered somewhere and [command spoofing](https://github.com/PowerShell/PowerShell/issues/17135) shouldn't be that easy)  and making use of the existing code as-is (by using a [proxy command](https://devblogs.microsoft.com/scripting/proxy-functions-spice-up-your-powershell-core-cmdlets/)) shouldn't illegal either.

In other words, why not building a *new* open source version of  `ConvertFrom-String` with [new features](https://github.com/PowerShell/PowerShell/issues/726#issuecomment-1002153246) that calls legacy Microsoft `Microsoft.PowerShell.Utility\ConvertFrom-String` cmdlet version for downwards compatibility (and legal reasons.) of the current features.

**Comment by iSazonov** [comment]
> In other words, why not building a _new_ open source version of `ConvertFrom-String` with [new features](https://github.com/PowerShell/PowerShell/issues/726#issuecomment-1002153246) that calls legacy Microsoft `Microsoft.PowerShell.Utility\ConvertFrom-String` cmdlet version for downwards compatibility (and legal reasons.) of the current features.

This already works with WinCompat feature.

**Comment by adamency** [comment]
@vexx32 
> I'm not sure that's really worth the effort. In every case where I've tried to use that cmdlet for anything more than the simplest tasks, I've ended up giving up after a good hour or so and just using regex instead.
A simple cmdlet to generate PSCustomObjects from strings via a regex pattern and match groups would be far more useful (if pretty straightforward) in my opinion.

I strongly disagree with this sentiment. `ConvertFrom-String` was precisely the most straightforward solution to convert unstructured legacy CLI output in order to use powershell cmdlets on it.

A very simple output is to simulate the equivalent of the `column -t` command in Unix.

With it in order to replace 

```
my_whitespace_separated_command_output | column -t
```

we simply had to do 

```
my_whitespace_separated_command_output | ConvertFrom-String | Format-Table -AutoSize
```

And for any delimiter:

to replace 

```
my_tab_separated_command_output | column -t -s $'\t'
```

we simply had to do 

```
my_tab_separated_command_output | ConvertFrom-String -Delimiter "`t" | Format-Table -AutoSize
```

Curently, even 7 years later in 2023, it's not possible anymore to have such a simple solution for this as the closest Powershell thing to this cmdlet would be `ConvertFrom-StringData` which only converts to a dicitionary, hence effectively allowing only two columns. With more than 2 columns, there seems to be simply no solution besides implementing the logic ourselves.

It's really unfortunate that such a simple and classic operation has to be so complicated as to need to make the calculations for the column sizes ourselves by iterating over all the fields of all the rows. We have to acknowledge that there will always be unstructured CLI output in our workflow like the CLI tools for any third party services like cloud services, package managers, password managers, web servers, etc..

`ConvertFrom-String` or an equivalent is one of the most important tools of a OOP shell as it servers as the bridge between the wild west and its own environment.
